### PR TITLE
BD-2645 Two Factor Authentication

### PR DESCRIPTION
**Pull request by trukhilio**
- 2FA in Settings
 - 2FA Modals UI

### CODE REVIEW

File: client/src/components/TwoFactorAuth/components/TwoFactorCodeModal.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      <div className={classes.codeModalContainer}>
+        <h3 className={classes.title}>Enter Backup Code</h3>
+        <p className={classes.subtitle}>
+          Sign into your account using the device you setup 2FA.
+        </p>
+        <p className={classes.message}>
+          Go to Settings &gt; 2 Factor Authentication to find the code if you do
+          not already have it recorded.
+        </p>
+        <form className={classes.form} onSubmit={handleSubmit}>
```
Comment (andreyvEze) [inline client/src/components/TwoFactorAuth/components/TwoFactorCodeModal.tsx:69]: could we create FinalForm for this?

File: client/src/pages/SettingsPage/components/TwoFactorAuth/TwoFactorAuth.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  successEnableModalSubtitle: 'Your 2FA has been successfully enabled.',
+};
+
+type Props = {
+  isTwoFAEnabled?: boolean;
+  loading: boolean;
+  genericErrorIsShowing: boolean;
+};
+
+const TwoFactorAuth: React.FC<Props> = props => {
```
Comment (andreyvEze) [inline client/src/pages/SettingsPage/components/TwoFactorAuth/TwoFactorAuth.tsx:33]: let's create 2 components for enable and disable

File: client/src/components/MainHeader/index.ts [hunk 3]
```diff
@@ -51,7 +54,7 @@ const actions = (dispatch: AppDispatch, ownProps: MainHeaderProps) => ({
         const { username } = getState(
           (state: ReduxState) => state.profile.lastAuthData,
         );
-        dispatch(clearCachedUser(username));
+        !hasTwoFactorAuth && dispatch(clearCachedUser(username));
```
Comment (andreyvEze) [inline client/src/components/MainHeader/index.ts:57]: `hasTwoFactorAuth` is selector not value, it is not working. you should use getState here like for `username` above

Comment (andreyvEze) [reply client/src/components/MainHeader/index.ts:57]: and let's refactor this - create saga for `resetLastAuthData` and use `clearCachedUser` there, so we have one place for this logic

File: client/src/api/edge.js [hunk 4]
```diff
@@ -223,4 +223,24 @@ export default class Edge {
       throw e;
     }
   }
+
+  async enableTwoFactorAuth(account) {
+    try {
+      await account.enableOtp();
+      return { status: 1 };
+    } catch (e) {
+      console.log(e);
```
Comment (andreyvEze) [inline client/src/api/edge.js:232]: console.error

File: client/src/api/edge.js [hunk 5]
```diff
@@ -223,4 +223,24 @@ export default class Edge {
...
+      throw e;
+    }
+  }
+
+  async disableTwoFactorAuth(account) {
+    try {
+      await account.disableOtp();
+      return { status: 1 };
+    } catch (e) {
+      console.log(e);
```
Comment (andreyvEze) [inline client/src/api/edge.js:242]: console.error

File: client/src/api/new-device-two-factor.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import Base from './base';
+
+export default class NewDeviceTwoFactor extends Base {
+  create({
```
Comment (andreyvEze) [inline client/src/api/new-device-two-factor.ts:4]: not sure we need to create new api, could we use `auth` for this?

File: client/src/components/LoginForm/LoginForm.tsx [hunk 7]
```diff
@@ -74,56 +96,103 @@ const LoginForm = (props: Props) => {
...
+  };
+
   const onForgotPassClose = () => {
     toggleForgotPass(false);
   };
 
   const exitPin = () => {
     resetLastAuthData();
-    clearCachedUser(lastAuthData.username);
+    !hasTwoFactorAuth && clearCachedUser(lastAuthData.username);
```
Comment (andreyvEze) [inline client/src/components/LoginForm/LoginForm.tsx:134]: will refactor this as we talked?

File: client/src/components/LoginForm/components/TwoFactorCodeModal.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+const TwoFactorCodeModal: React.FC<Props> = props => {
+  const { show, onClose, onSubmit, loading, otpError } = props;
+
+  const handleClose = () => {
+    !loading && onClose();
+  };
+
+  const RenderForm = (props: FormRenderProps<BackupFormValues>) => {
```
Comment (andreyvEze) [inline client/src/components/LoginForm/components/TwoFactorCodeModal.tsx:38]: could we create separate component for this?

File: client/src/redux/edge/actions.ts [hunk 9]
```diff
@@ -344,3 +348,31 @@ export const CLEAR_RECOVERY_RESULTS = `${prefix}/CLEAR_RECOVERY_RESULTS`;
...
+
+export const disableTwoFactor = (account: EdgeAccount) => ({
+  types: [
+    DISABLE_TWO_FACTOR_REQUEST,
+    DISABLE_TWO_FACTOR_SUCCESS,
+    DISABLE_TWO_FACTOR_FAILURE,
+  ],
+  promise: (api: Api) =>
+    minWaitTimeFunction(() => api.edge.disableTwoFactorAuth(account), 2000),
+});
```
Comment (andreyvEze) [inline client/src/redux/edge/actions.ts:378]: could we remove this and create action like `setHasTwoFactorAuth` intsead, let's discuss

File: client/src/components/TwoFactorApproveModal/TwoFactorApproveModal.tsx [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+          >
+            Not Me, Deny All
+          </button>
+        </div>
+      </Modal>
+    </>
+  );
+};
+
+export default TwoFactorAuth;
```
Comment (andreyvEze) [inline client/src/components/TwoFactorApproveModal/TwoFactorApproveModal.tsx:229]: let's refactor this file a little, for example move actual modal to new file

File: client/src/components/TwoFactorApproveModal/index.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+import { user } from '../../redux/profile/selectors';
+
+import TwoFactorApproveModal from './TwoFactorApproveModal';
+
+const reduxConnect = connect(
+  createStructuredSelector({
+    user,
+  }),
+  {},
```
Comment (andreyvEze) [inline client/src/components/TwoFactorApproveModal/index.ts:14]: remove second arg

File: client/src/pages/SettingsPage/components/TwoFactorAuth/TwoFactorAuth.tsx [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+  return (
+    <SecurityItem {...securityItemProps}>
+      <EdgeConfirmAction {...edgeConfirmActionProps} />
+      <DangerModal {...dangerModalProps} />
+      <SuccessModal {...successModalProps} />
+    </SecurityItem>
+  );
+};
+
+export default TwoFactorAuth;
```
Comment (andreyvEze) [inline client/src/pages/SettingsPage/components/TwoFactorAuth/TwoFactorAuth.tsx:192]: want to discuss how we could separate this in two files

File: client/src/redux/edge/actions.ts [hunk 13]
```diff
@@ -344,3 +348,10 @@ export const CLEAR_RECOVERY_RESULTS = `${prefix}/CLEAR_RECOVERY_RESULTS`;
 export const clearRecoveryResults = () => ({
   type: CLEAR_RECOVERY_RESULTS,
 });
+
+export const TOGGLE_TWO_FACTOR_AUTH = `${prefix}/TOGGLE_TWO_FACTOR_AUTH`;
+
+export const toggleTwoFactorAuth = (flag: boolean) => ({
```
Comment (andreyvEze) [inline client/src/redux/edge/actions.ts:354]: rename to `enabled`

File: client/src/api/middleware/edge.ts [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import { store } from '../../redux/init';
+
+export const HandleClearCachedUser = (
```
Comment (andreyvEze) [inline client/src/api/middleware/edge.ts:3]: rename to `handleClearCachedUser`

File: client/src/components/TwoFactorApproveModal/TwoFactorApproveModal.tsx [hunk 15]
```diff
@@ -92,9 +72,8 @@ const TwoFactorAuth: React.FC<Props> = props => {
         );
       }
       if (type === CLICK_TYPE.DENY_ALL) {
-        const { rejectVoucher } = edgeAccount;
         const promises = newDevicesList.map(async device => [
-          await rejectVoucher(device.voucherId),
+          await edgeAccount.rejectVoucher(device.voucherId),
```
Comment (andreyvEze) [inline client/src/components/TwoFactorApproveModal/TwoFactorApproveModal.tsx:76]: remove awaits

File: client/src/pages/SettingsPage/components/TwoFactorAuth/types.d.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    processing: boolean;
+    hideProcessing: boolean;
+  };
+  genericErrorIsShowing?: boolean;
+  loading: boolean;
+  processing: boolean;
+  toggleTwoFactorAuth: (enabled: boolean) => void;
+  showGenericErrorModal: () => void;
+  toggleLoading: (flag: boolean) => void;
+  setProcessing: (flag: boolean) => void;
```
Comment (andreyvEze) [inline client/src/pages/SettingsPage/components/TwoFactorAuth/types.d.ts:22]: please rename all `flag`s to named props depend on context - isLoading/isProcessing for example

File: client/src/util/login.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    login({ ...loginParams, voucherId });
+    return;
+  }
+  autoLogin({ voucherId, timerRef, loginParams, login });
+};
+
+export const autoLogin = (params: LoginProps) => {
+  const { timerRef } = params;
+  timerRef.current = setTimeout(
+    async () => await handleAutoLogin({ ...params }),
```
Comment (andreyvEze) [inline client/src/util/login.ts:53]: remove async await

File: client/src/components/TwoFactorApproveModal/TwoFactorApproveModal.tsx [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        );
+        const deviceData = { ...pendingVouchers };
+
+        if (pendingVouchers && pendingVouchers.deviceDescription) {
+          deviceData.deviceDescription = pendingVouchers.deviceDescription;
+        } else {
+          deviceData.deviceDescription = device.deviceDescription || 'N/A';
+        }
+
+        return deviceData;
```
Comment (andreyvEze) [inline client/src/components/TwoFactorApproveModal/TwoFactorApproveModal.tsx:54]: why do we return data here?

File: client/src/components/TwoFactorApproveModal/components/ModalComponent.tsx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+                  value={commonFormatTime(created)}
+                />
+
+                <h5 className={classes.denied}>
+                  Unless denied, access will be granted on
+                </h5>
+                <p className={classes.deniedValue}>
+                  {commonFormatTime(activates)}
+                </p>
+                <button
```
Comment (andreyvEze) [inline client/src/components/TwoFactorApproveModal/components/ModalComponent.tsx:75]: is it similar to <ActionButton /> ?

File: client/src/redux/edge/selectors.ts [hunk 20]
```diff
@@ -36,3 +36,8 @@ export const usersRecoveryQuestions = (state: ReduxState) =>
   state[prefix].usersRecoveryQuestions;
 export const recoveryAccountResults = (state: ReduxState) =>
   state[prefix].recoveryAccountResults;
+
+export const hasTwoFactorAuth = (state: ReduxState) =>
+  state[prefix].hasTwoFactorAuth;
+export const twoFactorAuthLoading = (state: ReduxState) =>
```
Comment (andreyvEze) [inline client/src/redux/edge/selectors.ts:42]: remove

File: client/src/redux/store.js [hunk 21]
```diff
@@ -41,6 +41,9 @@ export default function configureStore(api, history) {
           lastAuthData: store.getState().profile.lastAuthData,
           lastActivityDate: store.getState().profile.lastActivityDate,
         },
+        edge: {
+          hasTwoFactorAuth: store.getState().edge.hasTwoFactorAuth,
```
Comment (andreyvEze) [inline client/src/redux/store.js:45]: should we reset it on logout?

File: server/src/services/users/Info.mjs [hunk 22]
```diff
@@ -37,6 +38,15 @@ export default class UsersInfo extends Base {
       userObj.secretSetNotification = !!secretSetNotification;
     }
 
+    if (userObj.newDeviceTwoFactor.length > 0) {
+      const newDevicesValid = [];
+      for (const newDeviceItem of userObj.newDeviceTwoFactor) {
+        const isExpired = await NewDeviceTwoFactor.isExpired(newDeviceItem);
+        if (!isExpired) newDevicesValid.push(newDeviceItem);
+        userObj.newDeviceTwoFactor = newDevicesValid;
```
Comment (andreyvEze) [inline server/src/services/users/Info.mjs:46]: move outside of `for`

File: server/src/models/NewDeviceTwoFactor.mjs [hunk 23]
```diff
@@ -45,4 +47,17 @@ export class NewDeviceTwoFactor extends Base {
       createdAt,
     };
   }
+
+  static async isExpired(newDeviceItem) {
+    const createdTime =
+      new Date(newDeviceItem.createdAt).getTime() +
+      EXPIRED_NEW_DEVICE_DAYS * 24 * 60 * 60 * 1000;
```
Comment (andreyvEze) [inline server/src/models/NewDeviceTwoFactor.mjs:54]: set `EXPIRED_NEW_DEVICE_DAYS` as 7 * 24 * 60 * 60 * 1000

File: server/src/services/twoFactor/Reject.mjs [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+export default class NewDeviceTwoFactorReject extends Base {
+  static get validationRules() {
+    return {
+      voucherId: ['required', 'string'],
+    };
+  }
+  async execute({ voucherId }) {
+    const newDeviceTwoFactor = await NewDeviceTwoFactor.getItem({
+      voucherId,
+      status: NewDeviceTwoFactor.STATUS.REQUESTED,
```
Comment (andreyvEze) [inline server/src/services/twoFactor/Reject.mjs:15]: add userId

File: client/src/util/login.ts [hunk 25]
```diff
@@ -17,13 +17,15 @@ type LoginProps = {
...
   }) => void;
+  onSuccess: () => void;
 };
 
 const handleAutoLogin = async ({
   voucherId,
   timerRef,
   loginParams,
   login,
+  onSuccess,
```
Comment (andreyvEze) [inline client/src/util/login.ts:28]: rename

File: client/src/util/login.ts [hunk 26]
```diff
@@ -38,12 +40,24 @@ const handleAutoLogin = async ({
     return true;
   };
   const hasReadyVoucher = await voucherCheck();
+  const isRejected = await apis.auth.checkRejected(voucherId);
+
+  if (isRejected) {
+    try {
+      await apis.auth.deleteNewDeviceRequest(voucherId);
```
Comment (andreyvEze) [inline client/src/util/login.ts:47]: remove this, but test before

File: client/src/util/login.ts [hunk 27]
```diff
@@ -17,13 +17,15 @@ type LoginProps = {
     voucherId?: string;
     options?: { otpKey?: string };
   }) => void;
+  onSuccess: () => void;
```
Comment (andreyvEze) [inline client/src/util/login.ts:20]: * rename to AutoLoginProps
* loginParams to autoLoginParams and create type
* remove `options` and set voucherId not optional

File: client/src/util/login.ts [hunk 28]
```diff
@@ -17,13 +17,15 @@ type LoginProps = {
...
 
 const handleAutoLogin = async ({
   voucherId,
   timerRef,
   loginParams,
   login,
+  onSuccess,
 }: LoginProps) => {
   const voucherCheck = async () => {
     if (voucherId == null) return false;
```
Comment (andreyvEze) [inline client/src/util/login.ts:31]: remove

File: client/src/components/LoginForm/LoginForm.tsx [hunk 29]
```diff
@@ -102,6 +103,7 @@ const LoginForm = (props: Props) => {
     const isOtpError = edgeLoginFailure && edgeLoginFailure.reason === 'otp';
 
     if (isOtpError) {
+      setVoucherData(edgeLoginFailure.voucherActivates);
```
Comment (andreyvEze) [inline client/src/components/LoginForm/LoginForm.tsx:106]: setVoucherDate

File: client/src/components/LoginForm/LoginForm.tsx [hunk 30]
```diff
@@ -125,6 +133,8 @@ const LoginForm = (props: Props) => {
     });
   };
 
+  const onSuccessLogin = () => toggleBlockmodal(false);
```
Comment (andreyvEze) [inline client/src/components/LoginForm/LoginForm.tsx:136]: rename

File: client/src/util/login.ts [hunk 31]
```diff
@@ -38,12 +40,24 @@ const handleAutoLogin = async ({
     return true;
   };
   const hasReadyVoucher = await voucherCheck();
+  const isRejected = await apis.auth.checkRejected(voucherId);
+
+  if (isRejected) {
```
Comment (andreyvEze) [inline client/src/util/login.ts:45]: do it if `onSuccessLogin === true`

File: client/src/redux/profile/sagas.js [hunk 32]
```diff
@@ -32,15 +32,15 @@ export function* loginSuccess(history, api) {
     const wallets = yield select(fioWallets);
     api.client.setToken(action.data.jwt);
     if (wallets && wallets.length) yield put(setWallets(wallets));
+    if (action.otpKey || action.voucherId)
+      api.auth.deleteNewDeviceRequest(action.voucherId);
```
Comment (andreyvEze) [inline client/src/redux/profile/sagas.js:36]: try catch

File: server/src/router.mjs [hunk 33]
```diff
@@ -10,7 +10,16 @@ router.post('/auth', routes.auth.create);
 router.get('/auth/nonce', routes.auth.nonce);
 router.get('/auth/username/:email', routes.auth.username);
 router.post('/auth/new-device-two-factor', routes.newDeviceTwoFactor.create);
-router.delete('/auth/new-device-two-factor', checkAuth, routes.newDeviceTwoFactor.delete);
+router.post(
+  '/auth/new-device-two-factor/reject',
```
Comment (andreyvEze) [inline server/src/router.mjs:14]: rename to `update`

File: server/src/models/NewDeviceTwoFactor.mjs [hunk 34]
```diff
@@ -9,13 +9,25 @@ const { DataTypes: DT } = Sequelize;
...
+
   static init(sequelize) {
     super.init(
       {
         id: { type: DT.BIGINT, primaryKey: true, autoIncrement: true },
         voucherId: { type: DT.STRING, allowNull: false },
         deviceDescription: { type: DT.STRING, allowNull: true },
         userId: { type: DT.UUID },
+        status: {
+          type: DT.ENUM,
```
Comment (andreyvEze) [inline server/src/models/NewDeviceTwoFactor.mjs:27]: let's make is STRING

File: server/src/migrations/20211203211232-create-new-device-two-factor-table.js [hunk 35]
```diff
@@ -20,6 +20,7 @@ module.exports = {
       },
       voucherId: { type: DT.STRING, unique: true },
       deviceDescription: { type: DT.STRING },
+      status: { type: DT.STRING },
```
Comment (andreyvEze) [inline server/src/migrations/20211203211232-create-new-device-two-factor-table.js:23]: add default value

**Review by andreyvEze — COMMENTED** [review]
check comments

File: client/src/components/TwoFactorApproveModal/constants.ts [hunk 36]
```diff
@@ -0,0 +1,9 @@
+export const CLICK_TYPE = {
```
Comment (andreyvEze) [inline client/src/components/TwoFactorApproveModal/constants.ts:1]: let's rename to ACTION_TYPE

File: client/src/redux/profile/sagas.js [hunk 37]
```diff
@@ -32,9 +32,16 @@ export function* loginSuccess(history, api) {
     const wallets = yield select(fioWallets);
     api.client.setToken(action.data.jwt);
     if (wallets && wallets.length) yield put(setWallets(wallets));
-    if (action.otpKey || action.voucherId)
-      api.auth.deleteNewDeviceRequest(action.voucherId);
-    yield put(loadProfile());
+    if ((action.otpKey && action.voucherId) || action.voucherId) {
+      try {
+        const res = yield api.auth.deleteNewDeviceRequest(action.voucherId);
+        if (res) yield put(loadProfile());
```
Comment (andreyvEze) [inline client/src/redux/profile/sagas.js:38]: is there a case when we don;t need to loadProfile?

File: server/src/services/twoFactor/Update.mjs [hunk 38]
```diff
@@ -3,16 +3,29 @@ import X from '../Exception';
...
-export default class NewDeviceTwoFactorReject extends Base {
+export default class NewDeviceTwoFactorUpdate extends Base {
   static get validationRules() {
     return {
-      voucherId: ['required', 'string'],
+      data: [
+        'required',
+        {
+          nested_object: {
+            updateParams: {
```
Comment (andreyvEze) [inline server/src/services/twoFactor/Update.mjs:13]: i think we don't need to created `updateParams`, just props in it

File: client/src/redux/profile/sagas.js (hunk 37 above)
Comment (trukhilio) [reply client/src/redux/profile/sagas.js:38]: We have to wait until voucher will be deleted. In other case profile will be returned non updated.

File: server/src/services/twoFactor/Update.mjs [hunk 39]
```diff
@@ -3,16 +3,29 @@ import X from '../Exception';
...
+              nested_object: {
+                status: ['string'],
+                deviceDescription: ['string'],
```
Comment (andreyvEze) [inline server/src/services/twoFactor/Update.mjs:16]: check validation for additional params

File: server/src/router.mjs [hunk 40]
```diff
@@ -11,11 +11,11 @@ router.get('/auth/nonce', routes.auth.nonce);
 router.get('/auth/username/:email', routes.auth.username);
 router.post('/auth/new-device-two-factor', routes.newDeviceTwoFactor.create);
 router.post(
-  '/auth/new-device-two-factor/reject',
+  '/auth/new-device-two-factor/update',
```
Comment (andreyvEze) [inline server/src/router.mjs:14]: `/auth/new-device-two-factor/update/:voucherId`

**Pull request merged**